[samuelson] Fixing the lecture

bishmaybarik opened the pull request:
This PR brings the following changes to the lecture:

- [x] Fixes #530 : I have rectified the error in the mathematical equation
- [x] Fixes #456 : I have updated the lectures to use $\alpha$ and $\beta$ throughout; and this is consistent with the explanation on text as well
- [x] Fixes #531 : Making some more changes to follow John's suggestions
  - [x] large code cells with lots of plotting code are typically hidden now -- see, e.g., `def param_plot()`
  - [x] we might want to truncate the number of decimal points in output such as `(1.5371322893124, -0.9024999999999999)`
  - [x] we shouldn't have comments in cells, such as `### Test the categorize_solution function` -- if necessary, those comments should be ordinary markdown above the cells.
  - [x] I don't think we should have `plt.rcParams["figure.figsize"] = (11, 5)  #set default figure size`
  - [x] the align syntax after "Notice that" and in other places seems nonstandard

Additional tasks

- [x] run `black` over code-cells for PEP8 / code style compliance (line-length=80)
- [x] adjusted **bold** to *emphasis* for non-definitions in text (style-guide)

bishmaybarik commented:
I am not sure if I should completely delete the following three lines of code:

```python
### code to reverse-engineer a cycle
### y_t = r^t (c_1 cos(ϕ t) + c2 sin(ϕ t))
###
```
from this entire code chunk (it can be found [here](https://python.quantecon.org/samuelson.html#reverse-engineering-parameters-to-generate-damped-cycles)):
```python
### code to reverse-engineer a cycle
### y_t = r^t (c_1 cos(ϕ t) + c2 sin(ϕ t))
###

def f(r, ϕ):
    """
    Takes modulus r and angle ϕ of complex number r exp(j ϕ)
    and creates ρ1 and ρ2 of characteristic polynomial for which
    r exp(j ϕ) and r exp(- j ϕ) are complex roots.

    Returns the multiplier coefficient a and the accelerator coefficient b
    that verifies those roots.
    """
    g1 = cmath.rect(r, ϕ)  # Generate two complex roots
    g2 = cmath.rect(r, -ϕ)
    ρ1 = g1 + g2           # Implied ρ1, ρ2
    ρ2 = -g1 * g2
    b = -ρ2                # Reverse-engineer a and b that validate these
    a = ρ1 - b
    return ρ1, ρ2, a, b

## Now let's use the function in an example
## Here are the example parameters

r = .95
period = 10                # Length of cycle in units of time
ϕ = 2 * math.pi/period

## Apply the function

ρ1, ρ2, a, b = f(r, ϕ)

print(f"a, b = {a}, {b}")
print(f"ρ1, ρ2 = {ρ1}, {ρ2}")
```
Also, there are comments inside the code cell, like:
```python
## Now let's use the function in an example
## Here are the example parameters
```
which I think might be a little confusing if people miss out reading each line of the code. 

---
I plan to rewrite the entire thing like this:
```python
def f(r, ϕ):
    """
    Takes modulus r and angle ϕ of complex number r exp(j ϕ)
    and creates ρ1 and ρ2 of characteristic polynomial for which
    r exp(j ϕ) and r exp(- j ϕ) are complex roots.

    Returns the multiplier coefficient a and the accelerator coefficient b
    that verifies those roots.
    """
    g1 = cmath.rect(r, ϕ)  # Generate two complex roots
    g2 = cmath.rect(r, -ϕ)
    ρ1 = g1 + g2           # Implied ρ1, ρ2
    ρ2 = -g1 * g2
    b = -ρ2                # Reverse-engineer a and b that validate these
    a = ρ1 - b
    return ρ1, ρ2, a, b
```
Now let's use the function in an example. Here are the example parameters:
```python
r = .95
period = 10                # Length of cycle in units of time
ϕ = 2 * math.pi/period

## Apply the function
ρ1, ρ2, a, b = f(r, ϕ)

print(f"a, b = {a}, {b}")
print(f"ρ1, ρ2 = {ρ1}, {ρ2}")
```
---
Please let me know your thoughts on this @HumphreyYang and @mmcky. Once you give your inputs, I can make the changes accordingly.

mmcky commented:
thanks @bishmaybarik -- we should got he other way in the unicode aspects of this update. Let's add them back into the code (to keep them close to the math) and then work through the text and update the `a` and `b` references there to use unicode characters.

bishmaybarik commented:
> thanks @bishmaybarik -- we should got he other way in the unicode aspects of this update. Let's add them back into the code (to keep them close to the math) and then work through the text and update the `a` and `b` references there to use unicode characters.

Hi @mmcky, thanks for the suggestion. I'll replace `a` and `b` with the unicode characters $\alpha$ and $\beta$ in the the entire lecture then.

bishmaybarik commented:
Hi @mmcky , I have updated the entire lecture. I have replaced $a$ and $b$ with $\alpha$ and $\beta$ both in text as well as the codes. 

I have carefully gone through this as well as the results it generates, and it matches with the results we see in the original lecture. But it would be nice if you could review it whenever you have the time :-)

Please let me know if you would like to see any other changes.

mmcky commented on lectures/samuelson.md:
```diff
@@ -118,10 +118,10 @@ Let's assume that
-  consumption function $C_t = a Y_{t-1} + \gamma$.
-- $b$ is the "accelerator coefficient" in the "investment
-  accelerator" $I_t = b (Y_{t-1} - Y_{t-2})$.
+- $⍺$ is the marginal propensity to consume in the Keynesian
```
@bishmaybarik inside of `$` and `$$` we should use `\alpha` for latex to pick this up and render it rather than `utf8`?

bishmaybarik commented:
Hi @mmcky , I am sorry I misunderstood the assignment earlier. It was my bad. I'll quickly make the changes and update you about the new version when it's done. 

Thanks again!

HumphreyYang commented:
Hi @bishmaybarik and @mmcky,

I have made some big changes to make the code in the lecture cleaner in https://github.com/QuantEcon/lecture-python.myst/pull/561/commits/5c8f323ef9cc2043b69aada6b5a52a9997af36e6.

Some main changes are:

* Added `analyze_roots` and unified simulator `simulate_samuelson` so we can remove repetition later in the lecture.
* Simplified `categorize_solution` logic/messages.
* Cleaned up plotting.
* `Samuelson` class now uses analysis/simulator with legend/param fixes to remove repetition in the old code.
* Use unicode in the class and reduce the length of code to below 80 characters.

Please let me know what you think!

bishmaybarik commented:
> > thanks @HumphreyYang -- these are big changes and lots of reorganisation.
> > @bishmaybarik would you mind to read the updated version and let me know your thoughts on these changes?
> 
> Thanks for making such significant changes @HumphreyYang . Sure @mmcky , I am reviewing it right now and will let you know my thoughts on this as soon as possible.

These are nice changes @HumphreyYang !

@mmcky I have gone through the entire lecture, and these are some of my quick thoughts:
- The codes are now concise and to the point. 
- Figures look much better. There were some alignment issues in the legends, and Humphrey's version fixes that as well. 
- Reorganisation of some parts of the lecture makes the entire lecture easier to follow. 

Additionally, I reviewed the mathematical calculations and verified the results produced by the code. The math appears to be correct, and upon comparing this version with the previous one, I found that the outputs are consistent.

P.S. This is a great way to write codes -- I’m learning a lot from @HumphreyYang!

jstac commented:
Many thanks @bishmaybarik @HumphreyYang .  Very nice work.
